### ISSUE TITLE
Error to run Magiori FCN

### ISSUE THREAD

**Issue by resnow92** [remicres/otbtf#20]
Hi, 
I tried to run the **create_savedmodel_maggiori17_fullyconv.py** to train a model with my 15 bands images but I obtained this error:
![Capture](https://user-images.githubusercontent.com/37361708/80727951-a99e2600-8b06-11ea-964f-c5549cebb8e0.PNG)

I specified the number of channels (15). I tried with the **create_savedmodel_simple_fcn.py** and all is running well.

Thanks for your help.

**Comment by remicres** [comment]
There was an error in the documentation: wrong patches sizes during training step (I just corrected it in https://github.com/remicres/otbtf/commit/3daef29fa0a741497d47abc13b83829434ae511f). 
You must change the patches sizes for input and output, because the Maggiori model inputs patches of size 80x80 and produce output labels of size 16x16. To generate the two sets of patches, you can use the `PatchExtraction` application with `OTB_TF_NSOURCES=2`

**Comment by resnow92** [comment]
Thanks for your answer.
Why i have to use OTB_TF_NSOURCES=2 ? 
Sorry if i don't understand, why we have to generate 2 set of patchs ? you mean the label data have to be 16*16 ?
Last question, can we use this script for 3 classes classifications ? or is it for binary classification only ?
Thanks for your time, I have a too small background in deep learning and I don't understand well how to change parameters and data format to be read by tensorflow.

**Comment by remicres** [comment]
Setting `OTB_TF_NSOURCES=2` enables to extract the two set of patches in one time. If one patch lies outside the image, it is rejected from the sampling, and that guarantee that patches are sampled simultaneously for all sources. If you don't set `OTB_TF_NSOURCES`, it's harder to guarantee that you have the same patches. Suppose you have 2 images to sample simultaneously, like one label image and one remote sensing image: if your images have not exactly the same extend, some patches may fall outside one image but not the other, and this would lead to a unpaired dataset. Since 5704527823d8d224dbc743e781b03a40003c469a, `PatchesExtraction` also allows to set a no-data value for each source: with that you can reject every set of patches that contains some undesired data in at least one source.

**Comment by resnow92** [comment]
Ok thanks I think I understood, I have nice results with create_savedmodel_simple_fcn.py but i wanted to try this one.

**Comment by resnow92** [comment]
Sorry, I tried what you said, so I generates patch extraction with NSOURCES=2. As a result I have as you mentioned, 2 set of patchs: 
- 80x80 satellite patchs
- 16x16 satellite patchs
And I have one label file. Labels are 1 pixel in width.

So my question is, which file to use in the ModelTrain function for the  training.source2.il parameter ? As my label are not 16x16 but 1 pixel width.
Or I made something wrong with the patch extraction ?
thanks

**Comment by remicres** [comment]
I admit that this `outlabel` parameter is confusing. Actually, the `outlabel` creates a 1x1 patch image that contains the field value of the input vector data (field name is set with the `field` parameter). It is convenient for patch-based classification (when your model outputs a single value for a given input patch, and when the class is stored in a vector data field). But as soon as you want to do something else using rasters (e.g. semantic segmentation, image-to-image translation, ...) and that you have output patches as reference (e.g. 16x16), the `outlabel` is not used anymore. 

In your case, you just have to use the 80x80 (for input) and the 16x16 (for labels) patches. Just throw away the 1x1 patches, not used at all by the Maggiori model. By the way, I think that `outlabel` is not mandatory parameter, but you still have to provide the `field` parameter: not very logic if `outlabel` is not used, I conceed!

There is a little room for improvement in `PatchesExtraction`. In the future major release of otbtf, In next release, it could be great to improve this app parameters.

**Comment by resnow92** [comment]
Hum OK, many thanks for this very helpful precision ! I think I have all elements I need. Indeed, it was not clear for me, but as I said I am not an expert :)

**Comment by resnow92** [comment]
I have an error, don't know why, here is the log.
I will explore this issue tomorrow.
![image](https://user-images.githubusercontent.com/37361708/81962031-6f0eb000-9613-11ea-892a-a3fc3822c922.png)

**Comment by remicres** [comment]
There is a typo in your command line: replace `is.training=true` with `is_training=true`

**Comment by resnow92** [comment]
new error concerning shapes of input images ?
![image](https://user-images.githubusercontent.com/37361708/82028299-bc336600-9695-11ea-9097-8e9ec4ae9a65.png)

**Comment by remicres** [comment]
I don't know. How many channels have your images? Do you have modified the model? If yes, can you copy/paste it on pastebin or something like that

**Comment by resnow92** [comment]
I have 15 channels and I put the option -- n_channels 15 when I ran the python script. I didn't change the model.

**Comment by remicres** [comment]
The maggiori model is used for binary semantic segmentation. You must provide to `y` some binary labels (i.e. `0` or `1` pixels). The paper: https://hal.inria.fr/hal-01350706/document

**Comment by resnow92** [comment]
Ah, i think is the thing i missed ! I don't have binary raster label, only label from shapefile.

**Comment by remicres** [comment]
You can use the OTB application `Rasterization` to rasterize your vector over your image (`otbcli_Rasterization -im yourimage.tif -vec yourvector.shp ... -out outputlabelimage.tif`)

**Comment by resnow92** [comment]
Nice, thanks ! I will try this :)

**Comment by remicres** [comment]
to be continued in #34
